File: Soal4/Soal4.py
class Waktu:
    __jam = 0
    __menit = 0
    __detik = 0

    def __init__(self, *args) :
          if (len(args) == 3) :
            self.__jam = int(args[0])
            self.__menit = int(args[1])
            self.__detik = int(args[2])
          
          elif(len(args) == 0) :
            self.__jam = int(0)
            self.__menit = int(0)
            self.__detik = int(0)

          else :
            print("inputan anda tidak valid") 

    def setJam(self, jam):
        self.__jam = int(jam)
    
    def getJam(self):
        return self.__jam
    
    def setMenit(self, menit):
        self.__menit = int(menit)
    
    def getMenit(self):
        return self.__menit
    
    def setDetik(self, detik):
        self.__detik = int(detik)
    
    def getDetik(self):
        return self.__detik
    
    def inputWaktu(self):
        self.__jam = int(input(" Masukkan jam : "))
        self.__menit = int(input(" Masukkan menit : "))
        self.__detik = int(input(" Masukkan detik : "))
    
    def getWaktu(self):
        nJam = ""
        nMenit = ""
        nDetik = ""
        if(self.__jam < 10):
            nJam = "0"
        if(self.__menit < 10):
            nMenit = "0"
        if(self.__detik < 10):
            nDetik = "0"
        return nJam + str(self.__jam) + ":" + nMenit + str(self.__menit) + ":" + nDetik + str(self.__detik)

    def totalDetik(self):
        hasil = (int(3600) * self.__jam) + (int(60) * self.__menit) + self.__detik
        return hasil
    
    def waktuTotal(self, dtk:int):
        self.__menit = int(dtk/60)
        self.__detik = int(dtk%60)
        self.__jam = int(self.__menit/60)
        self.__menit = int(self.__menit%60)
    
    def cariDurasi(self, akhir):
        temp = Waktu()
        detikAwal = self.totalDetik()
        detikAkhir = akhir.totalDetik()
        if(detikAkhir < detikAwal):
            detikAkhir += 86400
        detikHasil = detikAkhir - detikAwal
        temp.waktuTotal(detikHasil)

        return temp

class Parkir:
    __no = ""
    __jenis = ""
    __masuk = Waktu()
    __keluar = Waktu()

    def __init__(self):
        self.__no = ""
        self.__jenis = ""
        self.__masuk = Waktu(0, 0, 0)
        self.__keluar = Waktu(0, 0, 0)

    def setNoKendaraan(self,no):
        self.__no = no
    
    def getNoKendaraan(self):
        return self.__no
    
    def setJenis(self, jenis):
        self.__jenis = jenis

    def getJenis(self):
        return self.__jenis
    
    def setWaktuMasuk(self, masuk):
        self.__masuk = masuk
    
    def getWaktuMasuk(self):
        return self.__masuk
    
    def setWaktuKeluar(self, keluar):
        self.__keluar = keluar
    
    def getWaktuKeluar(self):
        return self.__keluar
    
    def inputKendaraan(self):
        self.__no = input("No Kendaraan : ")
        self.__jenis = input("Jenis Kendaraan (Mobil/Motor): ")
        print("Jam Masuk: ")
        self.__masuk.inputWaktu()
        print("Jam Keluar: ")
        self.__keluar.inputWaktu()
    
    def getLamaParkir(self):
        return self.__masuk.cariDurasi(self.__keluar)
    
    def getJamParkir(self):
        hasil = 0
        if(self.getLamaParkir().getMenit()>=10 or self.getLamaParkir().getJam() >= 1):
            hasil = self.getLamaParkir().getJam()
            if(self.getLamaParkir().getMenit()>0 or self.getLamaParkir().getDetik() > 0):
                hasil = hasil + 1
        
        return hasil
    
    def getBiayaParkir(self):
        biaya = 0
        if(self.__jenis == "Motor"):
            biaya = self.getJamParkir() * 2000
        elif(self.__jenis == "Mobil"):
            biaya = self.getJamParkir() * 3000
        
        return biaya

class banyakKendaraan:
    __banyak = 0
    __kendaraan = []

    def __init__(self, banyak):
        self.__banyak = banyak

    def input(self):
        i = 0
        while(i < self.__banyak):
            print("\nKendaraan ke-", i+1)
            data = Parkir()
            data.inputKendaraan()
            self.__kendaraan.append(data)
            i += 1

    def output(self):
        print("\n\n                            Rekapitulasi Biaya parkir PT Parkir Jaya\n")
        no = int(1)
        i = int(0)
        print("=======================================================================================================")
        print("No  |  No Kendaraan   |   Jenis   |  Jam Masuk  |  Jam keluar  |  Lama Parkir  |  Lama jam  |  Biaya ")
        print("=======================================================================================================")
        while(i < self.__banyak):
            print(
            no, "    ",
            self.__kendaraan[i].getNoKendaraan(), "        ", 
            self.__kendaraan[i].getJenis(), "     ", 
            self.__kendaraan[i].getWaktuMasuk().getWaktu(), "     ", 
            self.__kendaraan[i].getWaktuKeluar().getWaktu(), "     ",   
            self.__kendaraan[i].getLamaParkir().getWaktu(), "      ", 
            self.__kendaraan[i].getJamParkir(), "        ", 
            self.__kendaraan[i].getBiayaParkir())
            no += 1
            i += 1
        print("=======================================================================================================")
    
    def totalBiaya(self):
        hasil = float(0)
        i = int(0)
        while(i<self.__banyak):
            hasil = hasil + self.__kendaraan[i].getBiayaParkir()
            i = i + 1
        return hasil

banyak = int(input("Masukkan banyak kendaraan : "))
parkir = banyakKendaraan(banyak)
parkir.input()
parkir.output()
print("Total Biaya Parkir = ", parkir.totalBiaya())

File: Soal4/Soal4.java
/* Nama program : Tugas 8 UTS No. 4.java
Nama : Gilman Nahdi Al-fathir
NPM : 140810210060
Tanggal buat : 23-10-2022
Deskripsi : parkir
******************************************************/
import java.util.Scanner;

class Waktu{
    private int jam;
    private int menit;
    private int detik;

    Waktu(int jam, int menit, int detik){
        this.jam = jam;
        this.menit = menit;
        this.detik = detik;

    }
    Waktu(){
        this.jam = 0;
        this.menit = 0;
        this.detik = 0;
    }

    public void setJam(int jam){
        this.jam = jam;
    }

    public int getJam(){
        return this.jam;
    }

    public void setMenit(int menit){
        this.menit = menit;
    }

    public int getMenit(){
        return this.menit;
    }

    public void setDetik(int detik){
        this.detik = detik;
    }
    
    public int getDetik(){
        return this.detik;
    }

    public void inputJam(){
        Scanner sc = new Scanner(System.in);
        System.out.print(" Masukkan jam   : ");
        this.jam = sc.nextInt();
        System.out.print(" Masukkan menit : ");
        this.menit = sc.nextInt();
        System.out.print(" Masukkan detik : ");
        this.detik = sc.nextInt();
    }

    public String getWaktu(){
        String nJam = "";
        String nMenit = "";
        String nDetik = "";
        if(this.jam < 10){
            nJam = "0";
        }
        if(this.menit < 10){
            nMenit = "0";
        }
        if(this.detik < 10){
            nDetik = "0";
        }
        
        return nJam + this.jam + ":" + nMenit + this.menit + ":" + nDetik + this.detik;
    }

    public int totalDetik(){
        int hasil = (this.jam*3600) + (this.menit*60) + (this.detik);
        return hasil;
    }

    public void waktuTotal(int dtk){
        this.menit = dtk / 60;
        this.detik = dtk % 60;
        this.jam = this.menit / 60;
        this.menit = this.menit % 60;
    }

    public Waktu cariDurasi(Waktu akhir){
        Waktu temp =  new Waktu();
        int detikAwal = this.totalDetik();
        int detikAkhir = akhir.totalDetik();
        if(detikAkhir < detikAwal){
            detikAkhir += 86400;
        }
        int detikHasil = detikAkhir - detikAwal;
        temp.waktuTotal(detikHasil);

        return temp;
    }
}

class Parkir{
    private String no;
    private String jenis;
    private Waktu masuk = new Waktu();
    private Waktu keluar = new Waktu();

    Parkir(){
        this.no = "";
        this.jenis = "";
    }
    Parkir(String no, String jenis) {
        this.no = no;
        this.jenis = jenis;
    }

    public void setNoKendaraan(String no){
        this.no = no;
    }
    public String getNoKendaraan(){
        return this.no;
    }
    public void setJenis(String jenis){
        this.jenis = jenis;
    }
    public String getJenis(){
        return this.jenis;
    }
    public void setWaktumasuk(Waktu masuk){
        this.masuk = masuk;
    }
    public Waktu getWaktumasuk(){
        return this.masuk;
    }
    public void setWaktukeluar(Waktu keluar){
        this.keluar = keluar;
    }
    public Waktu getWaktukeluar(){
        return this.keluar;
    }

    public void inputKendaraan(){
        Scanner sc = new Scanner(System.in);
        System.out.print("No Kendaraan : ");
        this.no = sc.nextLine();
        System.out.print("Jenis Kendaraan (Mobil/Motor): ");
        this.jenis = sc.nextLine();
        System.out.println("Jam Masuk : ");
        this.masuk.inputJam();
        System.out.println("Jam Keluar : ");
        keluar.inputJam();
    }

    public Waktu getLamaParkir(){
        return this.masuk.cariDurasi(this.keluar);
    }

    public int getJamParkir(){
        int hasil = 0;
        if(this.getLamaParkir().getMenit() >= 10 || this.getLamaParkir().getJam() >= 1){
            hasil = this.getLamaParkir().getJam();
            if( this.getLamaParkir().getMenit() > 0 || this.getLamaParkir().getDetik() > 0){
                hasil += 1;
            }
        }
        return hasil;
    }

    public int getBiayaParkir(){
        int biaya = 0;
        if (this.jenis == "Motor" || this.jenis == "motor") {
            biaya = getJamParkir()*2000;
        } else if (this.jenis == "Mobil" || this.jenis == "mobil"){
            biaya = getJamParkir()*3000;
        } else {
            return 0;
        }
        return biaya;
    }    
}

class BanyakKendaraan{
    private int banyak;
    private Parkir kendaraan[] = new Parkir[50];

    BanyakKendaraan(int banyak){
        this.banyak = banyak;

        for(int i = 0; i < this.banyak; i++){
            kendaraan[i] = new Parkir();
        }
    }

    public void input(){
        for(int i = 0; i < this.banyak; i++){ 
            System.out.println("\nKendaraan ke-" + (i+1));
            this.kendaraan[i].inputKendaraan();
        }
    }

    public void output(){
        System.out.println("\n\n                            Rekapitulasi Biaya parkir PT Parkir Jaya\n");
            int no = 1;
            System.out.println("=======================================================================================================");
            System.out.println("No  |  No Kendaraan   |   Jenis   |  Jam Masuk  |  Jam keluar  |  Lama Parkir  |  Lama jam  |  Biaya ");
            System.out.println("=======================================================================================================");
            for(int i = 0; i < this.banyak; i++){
                System.out.println(
                    no + "      " +
                    this.kendaraan[i].getNoKendaraan() + "           " + 
                    this.kendaraan[i].getJenis()+ "      " + 
                    this.kendaraan[i].getWaktumasuk().getWaktu() + "      " + 
                    this.kendaraan[i].getWaktukeluar().getWaktu() + "      " +   
                    this.kendaraan[i].getLamaParkir().getWaktu() + "         " + 
                    this.kendaraan[i].getJamParkir() + "         " + 
                    this.kendaraan[i].getBiayaParkir());
                no++;
            }
            System.out.println("=======================================================================================================");
    }

    public float totalBiaya(){
        float hasil = 0;
        for(int i = 0; i < this.banyak; i++){
            hasil += this.kendaraan[i].getBiayaParkir();
        }
        return hasil;
    }
}


public class Soal4{
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Masukkan banyak kendaraan : ");
        int banyak = sc.nextInt();

        BanyakKendaraan parkir = new BanyakKendaraan(banyak);
        parkir.input();
        parkir.output();

        System.out.println("Total Biaya Parkir = " + parkir.totalBiaya());
    }
}
